File: acorn-core/src/test/java/com/ardley/acorn/permission/PermissionSetTest.java
package com.ardley.acorn.permission;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

/**
 * Tests PermissionSet JSON parsing including shorthand formats, granular configurations,
 * and error handling for malformed input.
 */
class PermissionSetTest {

    @Test
    @DisplayName("Parses 'allow all' shorthand into allowAll flag")
    void parseAllowAllShorthand() {
        PermissionSet ps = PermissionSet.fromJson("""
            {"allow": "all"}
            """);

        assertThat(ps.isAllowAll()).isTrue();
        assertThat(ps.hasAllowFor("anything")).isTrue();
        assertThat(ps.hasUnconditionalDeny("anything")).isFalse();
    }

    @Test
    @DisplayName("Parses granular allow and deny maps")
    void parseGranularPermissions() {
        PermissionSet ps = PermissionSet.fromJson("""
            {
                "allow": {
                    "ListUsers": "all",
                    "UpdateUser": {"department": {"match": "principal"}}
                },
                "deny": {
                    "DeleteUser": "all"
                }
            }
            """);

        assertThat(ps.isAllowAll()).isFalse();
        assertThat(ps.hasAllowFor("ListUsers")).isTrue();
        assertThat(ps.hasAllowFor("UpdateUser")).isTrue();
        assertThat(ps.hasAllowFor("CreateUser")).isFalse();
        assertThat(ps.hasUnconditionalDeny("DeleteUser")).isTrue();
        assertThat(ps.hasUnconditionalDeny("UpdateUser")).isFalse();
    }

    @Test
    @DisplayName("Allow level returns correct PermissionLevel type")
    void allowLevelTypes() {
        PermissionSet ps = PermissionSet.fromJson("""
            {
                "allow": {
                    "ListUsers": "all",
                    "UpdateUser": {"department": {"match": "principal"}},
                    "DeleteUser": "none"
                }
            }
            """);

        assertThat(ps.allowLevel("ListUsers")).isPresent().get().matches(PermissionLevel::isAll);
        assertThat(ps.allowLevel("UpdateUser")).isPresent().get().matches(PermissionLevel::isScoped);
        assertThat(ps.allowLevel("DeleteUser")).isPresent().get().matches(PermissionLevel::isNone);
        assertThat(ps.allowLevel("Unknown")).isEmpty();
    }

    @Test
    @DisplayName("Deny with allow all: deny entries override for specific actions")
    void denyWithAllowAll() {
        PermissionSet ps = PermissionSet.fromJson("""
            {
                "allow": "all",
                "deny": {"DeleteUser": "all", "SuspendUser": "all"}
            }
            """);

        assertThat(ps.isAllowAll()).isTrue();
        assertThat(ps.hasUnconditionalDeny("DeleteUser")).isTrue();
        assertThat(ps.hasUnconditionalDeny("SuspendUser")).isTrue();
        assertThat(ps.hasUnconditionalDeny("UpdateUser")).isFalse();
    }

    @Test
    @DisplayName("Empty JSON produces empty permission set that denies everything")
    void emptyJsonDeniesAll() {
        PermissionSet ps = PermissionSet.fromJson("{}");

        assertThat(ps.isAllowAll()).isFalse();
        assertThat(ps.hasAllowFor("anything")).isFalse();
    }

    @Test
    @DisplayName("Malformed JSON throws IllegalArgumentException")
    void malformedJsonThrows() {
        assertThatThrownBy(() -> PermissionSet.fromJson("not json"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("Invalid permission JSON");
    }

    @Test
    @DisplayName("Invalid permission level string throws")
    void invalidLevelStringThrows() {
        assertThatThrownBy(() -> PermissionSet.fromJson("""
            {"allow": {"ListUsers": "maybe"}}
            """))
                .isInstanceOf(IllegalArgumentException.class);
    }
}

File: node/packages/core/src/__tests__/permission-set.test.ts
import { describe, test, expect } from 'bun:test';
import { PermissionSet } from '../permission-set.js';

describe('PermissionSet', () => {
  test('allowAll returns superadmin set', () => {
    const perms = PermissionSet.allowAll();
    expect(perms.isAllowAll()).toBe(true);
    expect(perms.hasAllowFor('anything')).toBe(true);
  });

  test('empty returns deny-all set', () => {
    const perms = PermissionSet.empty();
    expect(perms.isAllowAll()).toBe(false);
    expect(perms.hasAllowFor('anything')).toBe(false);
  });

  test('parses superadmin shorthand', () => {
    const perms = PermissionSet.fromJson('{"allow": "all"}');
    expect(perms.isAllowAll()).toBe(true);
  });

  test('parses allow map with "all" level', () => {
    const perms = PermissionSet.fromJson('{"allow": {"ListUsers": "all"}}');
    expect(perms.hasAllowFor('ListUsers')).toBe(true);
    expect(perms.hasAllowFor('Other')).toBe(false);
    const level = perms.allowLevel('ListUsers');
    expect(level?.type).toBe('all');
  });

  test('parses scoped allow', () => {
    const perms = PermissionSet.fromJson(JSON.stringify({
      allow: { UpdateUser: { department: { match: 'principal' } } },
    }));
    const level = perms.allowLevel('UpdateUser');
    expect(level?.type).toBe('scoped');
  });

  test('parses deny map', () => {
    const perms = PermissionSet.fromJson(JSON.stringify({
      allow: { ListUsers: 'all' },
      deny: { DeleteUser: 'all' },
    }));
    expect(perms.hasUnconditionalDeny('DeleteUser')).toBe(true);
    expect(perms.hasUnconditionalDeny('ListUsers')).toBe(false);
  });

  test('scoped deny is not unconditional', () => {
    const perms = PermissionSet.fromJson(JSON.stringify({
      deny: { UpdateUser: { department: { equals: 'Executive' } } },
    }));
    expect(perms.hasUnconditionalDeny('UpdateUser')).toBe(false);
    expect(perms.denyLevel('UpdateUser')?.type).toBe('scoped');
  });

  test('throws on invalid JSON', () => {
    expect(() => PermissionSet.fromJson('not json')).toThrow('Invalid permission JSON');
  });

  test('throws on invalid level string', () => {
    expect(() => PermissionSet.fromJson('{"allow": {"X": "bogus"}}')).toThrow('Invalid permission level');
  });
});
